source/game_sa/CFont.cpp

```diff
@@ -56,206 +90,582 @@ void CFont::Initialise() {
+                    float u3 = u1;
+                    float v3 = v2 - 0.0021f;
+                    float u4 = u2;
+                    float v4 = v2 - 0.0021f;
+
+                    CSprite2d::AddToBuffer(rt, RenderState.m_color, u1, v1, u2, v2, u3, v3, u4, v4);
+                }
+                else {
+                    rt.bottom = 16.0f * RenderState.m_fHeight + y;
+
+                    float u1 = (character & 0xF) / 16.0f;
+                    float v1 = (character >> 4) / 12.8f + 0.0021f;
+                    float u2 = u1 + 0.0615f;
+                    float v2 = v1;
+                    float u3 = u1;
+                    float v3 = v2 - 0.016f;
+                    float u4 = u2;
+                    float v4 = v2 - 0.015f;
+
+                    CSprite2d::AddToBuffer(rt, RenderState.m_color, u1, v1, u2, v2, u3, v3, u4, v4);
+                }
+            }
+            else {
+                rt.top = y + 0.015f;
+                rt.right = 32.0f * RenderState.m_fWidth + x;
+                rt.bottom = 20.0f * RenderState.m_fHeight + y + 0.015f;
+
+                float u1 = (character & 0xF) / 16.0f;
+                float v1 = (character >> 4) / 12.8f + 0.00055f;
+                float u2 = u1 + 0.0615f;
+                float v2 = (character >> 4) / 12.8f + 0.078125f;
+                float u3 = u1;
+                float v3 = v2 - 0.016f;
+                float u4 = u2;
+                float v4 = v2 - 0.015f;
+
+                CSprite2d::AddToBuffer(rt, RenderState.m_color, u1, v1, u2, v2, u3, v3, u4, v4);
+            }
+        }
+    }
+    */
+}
+
+// 0x718F00
 char* CFont::ParseToken(char* text, CRGBA& color, bool isBlip, char* tag)
 {
-    return ((char* (__cdecl*)(char*, CRGBA&, bool, char*))0x718F00)(text, color, isBlip, tag);
-}
-
+    // info about tokens: https://gtamods.com/wiki/GXT#Tokens
+
+    char* next = ++text;
+
+    auto ApplyStyle = [&](eHudColours hudColor) {
+        if (!isBlip)
+            color = HudColour.GetRGBA(hudColor, color.a);
+
+        if (tag)
+            *tag = *next;
+    };
+
+    switch (*next) {
+    case '<':
+        m_nExtraFontSymbolId = EXSYMBOL_DPAD_LEFT;
+        break;
+    case '>':
+        m_nExtraFontSymbolId = EXSYMBOL_DPAD_RIGHT;
+        break;
+    case 'A':
+    case 'a':
+        m_nExtraFontSymbolId = EXSYMBOL_L3;
+        break;
+    case 'B':
+    case 'b':
+        ApplyStyle(HUD_COLOUR_DARK_BLUE);
+        break;
+    case 'C':
+    case 'c':
+        m_nExtraFontSymbolId = EXSYMBOL_R3;
+        break;
+    case 'D':
+    case 'd':
+        m_nExtraFontSymbolId = EXSYMBOL_DPAD_DOWN;
+        break;
+    case 'G':
+    case 'g':
+        ApplyStyle(HUD_COLOUR_GREEN);
+        break;
+    case 'H':
+    case 'h':
+        if (!isBlip) {
+            color = {
+                    std::min<uint8>(color.r * 1.5f, 255.0f),
+                    std::min<uint8>(color.g * 1.5f, 255.0f),
+                    std::min<uint8>(color.b * 1.5f, 255.0f),
```

Comment:
Must use casts here, otherwise the compiler will.. complain.
```suggestion
                    (uint8)std::min((float)color.r * 1.5f, 255.0f),
                    (uint8)std::min((float)color.g * 1.5f, 255.0f),
                    (uint8)std::min((float)color.b * 1.5f, 255.0f),
```